service/history/historyReplicator.go

```diff
@@ -68,167 +77,238 @@ func newHistoryReplicator(shard ShardContext, historyEngine *historyEngineImpl,
+	incomingVersion := request.GetVersion()
+	replicationInfo := request.ReplicationInfo
+	rState := msBuilder.GetReplicationState()
+	if rState.LastWriteVersion > incomingVersion {
+		// Replication state is already on a higher version, we can drop this event
+		// TODO: We need to replay external events like signal to the new version
+		logger.Warnf("Dropping stale replication task. CurrentV: %v, LastWriteV: %v, LastWriteEvent: %v, IncomingV: %v.",
+			rState.CurrentVersion, rState.LastWriteVersion, rState.LastWriteEventID, incomingVersion)
+		return nil, nil
+	} else if rState.LastWriteVersion < incomingVersion {
+		// Check if this is the first event after failover
+		logger.Infof("First Event after replication. CurrentV: %v, LastWriteV: %v, LastWriteEvent: %v, IncomingV: %v.",
+			rState.CurrentVersion, rState.LastWriteVersion, rState.LastWriteEventID, incomingVersion)
+		previousActiveCluster := r.clusterMetadata.ClusterNameForFailoverVersion(rState.LastWriteVersion)
+		ri, ok := replicationInfo[previousActiveCluster]
+		if !ok {
+			logger.Errorf("No ReplicationInfo Found For Previous Active Cluster. Previous Active Cluster: %v, Request Source Cluster: %v, Request ReplicationInfo: %v.",
+				previousActiveCluster, request.GetSourceCluster(), request.ReplicationInfo)
+			// TODO: Handle missing replication information, #840
+			return nil, ErrMissingReplicationInfo
 		}
 
-		// Check for out of order replication task and store it in the buffer
-		if firstEvent.GetEventId() > msBuilder.GetNextEventID() {
-			logger.Debugf("Buffer out of order replication task.  NextEvent: %v, FirstEvent: %v",
-				msBuilder.GetNextEventID(), firstEvent.GetEventId())
-			r.metricsClient.IncCounter(metrics.ReplicateHistoryEventsScope, metrics.BufferedReplicationTaskCounter)
+		// Detect conflict
+		if ri.GetLastEventId() < rState.LastWriteEventID {
+			logger.Infof("Conflict detected.  State: {CurrentV: %v, LastWriteV: %v, LastWriteEvent: %v}, ReplicationInfo: {PrevActiveCluster: %v, V: %v, LastEventID: %v}",
+				rState.CurrentVersion, rState.LastWriteVersion, rState.LastWriteEventID,
+				previousActiveCluster, ri.GetVersion(), ri.GetLastEventId())
 
-			if err := msBuilder.BufferReplicationTask(request); err != nil {
-				logger.Errorf("Failed to buffer out of order replication task.  Err: %v", err)
-				return errors.New("failed to add buffered replication task")
+			resolver := r.getNewConflictResolver(context, logger)
+			msBuilder, err = resolver.reset(uuid.New(), ri.GetLastEventId(), msBuilder.GetExecutionInfo().StartTimestamp)
+			logger.Infof("Completed Resetting of workflow execution.  NextEventID: %v. Err: %v", msBuilder.GetNextEventID(), err)
+			if err != nil {
+				return nil, err
 			}
+		} else if ri.GetLastEventId() > rState.LastWriteEventID {
+			// TODO handle this case
+			logger.Errorf("Conflict detected.  State: {CurrentV: %v, LastWriteV: %v, LastWriteEvent: %v}, ReplicationInfo: {PrevActiveCluster: %v, V: %v, LastEventID: %v}",
```

Comment:
I don't think I get this.  Can you explain?

Reply:
say if we have a bug, this case can happen, we should return some unique error and do logging